dcore/state.py

```diff
@@ -71,24 +74,47 @@ def dump(self):
             self.output.dumplist = self.fieldlist
 
         funcs = self.xn.split()
+        field_dict = {name: func for (name, func) in zip(self.fieldlist, funcs)}
         to_dump = []
-        for name, f in zip(self.fieldlist, funcs):
+        for name, f in field_dict.iteritems():
             if name in self.output.dumplist:
                 to_dump.append(f)
             f.rename(name=name)
 
-        dumpdir = path.join("results", self.output.dirname)
+        if self.output.steady_state_dump_err:
+            init_funcs = self.x_init.split()
+            for name, f, f_init in zip(self.fieldlist, funcs, init_funcs):
+                if name in self.output.dumplist:
+                    err = Function(f.function_space(), name=name+'err').assign(f-f_init)
+                    field_dict[name+"err"] = err
+                    to_dump.append(err)
 
-        outfile = path.join(dumpdir, "field_output.pvd")
+        self.dumpdir = path.join("results", self.output.dirname)
+
+        outfile = path.join(self.dumpdir, "field_output.pvd")
         if self.dumpfile is None:
-            if path.exists(dumpdir):
-                exit("results directory '%s' already exists" % dumpdir)
+            if path.exists(self.dumpdir):
+                exit("results directory '%s' already exists" % self.dumpdir)
             self.dumpcount = itertools.count()
             self.dumpfile = File(outfile, project_output=self.output.project_fields)
 
+            if self.diagnostics is not None:
+                self.diagnostic_data = {}
+                for name in field_dict.keys():
+                    self.diagnostic_data[name] = {"l2":[]}
+
         if (next(self.dumpcount) % self.output.dumpfreq) == 0:
             self.dumpfile.write(*to_dump)
 
+            if self.diagnostics is not None:
```

Comment:
So I would do the following.  Have a `register` method on the Diagnostics class.  So now I do:

```python
diags = MyDiagnostics()
diags.register(names, of, fields)

state = State(..., diags=diags, ...)
```

Then in `dump`.

```python
for name in diag.fields:
    data = diag.l2(field_dict[name])
   diag_data[name]["l2"].append(data)
```